prov/rxd/src/rxd_cq.c

```diff
@@ -229,391 +239,523 @@ static void rxd_ep_recv_data(struct rxd_ep *ep, struct rxd_x_entry *rx_entry,
+{
+	struct rxd_x_entry *rx_entry;
+	struct rxd_op_pkt *pkt = (struct rxd_op_pkt *) (pkt_entry->pkt);
+
+	if (pkt->pkt_hdr.flags & RXD_RETRY) {
+		rx_entry = rxd_check_active(ep, pkt);
+		if (rx_entry)
+			goto ack;
+	}
+
+	if (pkt->pkt_hdr.seq_no !=
+	    rxd_get_next_no(ep->peers[pkt->pkt_hdr.peer].rx_seq_no)) {
+		if (rxd_env.ooo_rdm) {
+			rxd_remove_rx_pkt(ep, pkt_entry);
+			dlist_insert_order(&ep->peers[pkt->pkt_hdr.peer].buf_ops,
+					   &rxd_op_order, &pkt_entry->d_entry);
 			return;
+		}
+
+		if (ep->peers[pkt->pkt_hdr.peer].rx_seq_no !=
+		    ep->peers[pkt->pkt_hdr.peer].last_ack_seq_no)
+			goto ack;
+		goto release;
 	}
 
-	if (!(tx_entry->flags & RXD_NO_COMPLETION)) {
-		fastlock_acquire(&ep->util_ep.tx_cq->cq_lock);
-		tx_cq->write_fn(tx_cq, &tx_entry->cq_entry);
-		fastlock_release(&ep->util_ep.tx_cq->cq_lock);
-		if (cntr)
-			cntr->cntr_fid.ops->add(&cntr->cntr_fid, 1);
+	rx_entry = rxd_match_rx(ep, pkt_entry);
+	if (!rx_entry) {
+		rxd_remove_rx_pkt(ep, pkt_entry);
+		return;
 	}
 
-	rxd_tx_entry_free(ep, tx_entry);
+	fastlock_acquire(&ep->util_ep.rx_cq->cq_lock);
+	rxd_progress_op(ep, pkt_entry, rx_entry);
+
+	if (rxd_env.ooo_rdm && !dlist_empty(&ep->peers[pkt->pkt_hdr.peer].buf_ops))
+		rxd_progress_buf_ops(ep, pkt->pkt_hdr.peer);
+
+	fastlock_release(&ep->util_ep.rx_cq->cq_lock);
+
+ack:
+	rxd_ep_post_ack(ep, pkt->pkt_hdr.peer);
+release:
+	rxd_remove_rx_pkt(ep, pkt_entry);
+	rxd_release_repost_rx(ep, pkt_entry);
 }
 
-void rxd_handle_recv_comp(struct rxd_ep *ep, struct fi_cq_msg_entry *comp)
+static void rxd_handle_cts(struct rxd_ep *ep, struct fi_cq_msg_entry *comp,
+			   struct rxd_pkt_entry *pkt_entry)
 {
-	struct rxd_pkt_entry *pkt_entry, *pkt_entry_head;
-	struct rxd_ctrl_pkt *ctrl_pkt;
-	struct rxd_data_pkt *data_pkt;
-	struct slist_entry *item, *prev;
-	int ret = 0;
+	struct rxd_cts_pkt *cts = (struct rxd_cts_pkt *) (pkt_entry->pkt);
 
-	FI_DBG(&rxd_prov, FI_LOG_EP_CTRL, "got recv completion\n");
+	ep->peers[cts->peer_addr].peer_addr = cts->dg_addr;
 
-	pkt_entry = container_of(comp->op_context, struct rxd_pkt_entry, context);
-	ep->posted_bufs--;
+	rxd_update_peer(ep, cts->peer_addr, cts->dg_addr);
+}
 
-	if (rxd_is_ctrl_pkt(pkt_entry)) {
-		ctrl_pkt = rxd_get_ctrl_pkt(pkt_entry);
-		switch (ctrl_pkt->ctrl_hdr.type) {
-		case RXD_RTS:
-			ret = rxd_handle_rts(ep, comp, pkt_entry);
-			break;
-		case RXD_CTS:
-			rxd_handle_cts(ep, comp, ctrl_pkt);
-			break;
-		case RXD_ACK:
-			rxd_handle_ack(ep, comp, ctrl_pkt);
+static void rxd_handle_ack(struct rxd_ep *ep, struct fi_cq_msg_entry *comp,
+			   struct rxd_pkt_entry *ack_entry)
+{
+	struct rxd_x_entry *tx_entry;
+	struct rxd_ack_pkt *ack = (struct rxd_ack_pkt *) (ack_entry->pkt);
+	struct rxd_pkt_entry *pkt_entry;
+	struct rxd_cq *tx_cq = rxd_ep_tx_cq(ep);
+	struct util_cntr *cntr = ep->util_ep.tx_cntr;
+	fi_addr_t peer = ack->pkt_hdr.peer;
+	struct rxd_pkt_hdr *hdr;
+
+	for (; !dlist_empty(&ep->peers[peer].unacked);
+	     ep->peers[peer].free_unacked++) {
```

Comment:
I would convert this into a while() statement.  There's no for-loop initialization code, and the update code is not directly related to the check being made.